File: 2020/python/day3.py
#!/usr/bin/env python3
import sys
import functools

def day3(input_file):
    with open(input_file, "r") as f:
        input = f.read()

    print("Day 3: Toboggan Trajectory")
    print(f"\tPart 1: {__part1(input)}")
    print(f"\tPart 2: {__part2(input)}")

def __slope_trees(rows, slope):
        (r, d) = slope
        x = 0
        trees = 0
        for i in range(0, len(rows), d):
            if x >= len(rows[i]):
                x = x - len(rows[i])
            if rows[i][x] == "#":
                trees += 1
            x += r
        return trees

def __part1(input):
    return __slope_trees(input.splitlines(), (3,1))

def __part2(input):
    slopes = (
        (1,1),
        (3,1),
        (5,1),
        (7,1),
        (1,2),
    )

    return functools.reduce(lambda acc, cur: acc * __slope_trees(input.splitlines(), cur), slopes, 1)


if __name__ == "__main__":
    day3(sys.argv[1])

File: 2020/javascript/day3.js
#!/usr/bin/env node

const fs = require("fs");

const slope_trees = (rows, slope) => {
  const [r, d] = slope;
  let x = 0;
  let trees = 0;
  for (let i = 0; i < rows.length; i += d) {
    if (x >= rows[i].length) {
      x = x - rows[i].length;
    }
    if (rows[i][x] === "#") {
      trees++;
    }
    x += r;
  }
  return trees;
};
const part1 = (input) => slope_trees(input.split("\n"), [3, 1]);

const part2 = (input) =>
  [
    [1, 1],
    [3, 1],
    [5, 1],
    [7, 1],
    [1, 2],
  ].reduce((acc, cur) => acc * slope_trees(input.split("\n"), cur), 1);

const day3 = (inputFile) => {
  const input = fs.readFileSync(inputFile, "utf8");

  console.log("Day 3: Toboggan Trajectory");
  console.log(`\tPart 1: ${part1(input)}`);
  console.log(`\tPart 2: ${part2(input)}`);
};

module.exports = day3;

if (require.main === module) day3(process.argv[2]);
